crates/sui-axelar-cgp/move/sources/channel.move

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+module axelar::channel {
+    use axelar::axelar;
+    use axelar::axelar::Axelar;
+    use axelar::message;
+    use axelar::message::Message;
+    use sui::bcs;
+    use sui::object;
+    use sui::object::UID;
+    use sui::tx_context::TxContext;
+    use sui::vec_set;
+    use sui::vec_set::VecSet;
+
+    /// Generic target for the messaging system.
+    ///
+    /// This struct is required on the Sui side to be the destination for the
+    /// messages sent from other chains. Even though it has a UID field, it does
+    /// not have a `key` ability to force wrapping.
+    ///
+    /// Notes:
+    ///
+    /// - Does not have key to prevent 99% of the mistakes related to access management.
+    /// Also prevents arbitrary Message destruction if the object is shared. Lastly,
+    /// when shared, `Channel` cannot be destroyed, and its contents will remain locked
+    /// forever.
+    ///
+    /// - Allows asset or capability-locking inside. Some applications might
+    /// authorize admin actions through the bridge (eg by locking some `AdminCap`
+    /// inside and getting a `&mut AdminCap` in the `consume_message`);
+    ///
+    /// - Can be destroyed freely as the `UID` is guaranteed to be unique across
+    /// the system. Destroying a channel would mean the end of the Channel cycle
+    /// and all further messages will have to target a new Channel if there is one.
+    ///
+    /// - Does not contain direct link to the state in Sui, as some functions
+    /// might not take any specific data (eg allow users to create new objects).
+    /// If specific object on Sui is targeted by this `Channel`, its reference
+    /// should be implemented using the `data` field.
+    ///
+    /// - The funniest and extremely simple implementation would be a `Channel<ID>`
+    /// since it actually contains the data required to point at the object in Sui.
+
+    /// For when trying to consume the wrong object.
+    const EWrongDestination: u64 = 0;
+    /// For when message has already been processed and submitted twice.
+    const EDuplicateMessage: u64 = 2;
+
+    struct Channel<T: store> has store {
+        /// Unique ID of the target object which allows message targeting
+        /// by comparing against `id_bytes`.
+        id: UID,
+        /// Messages processed by this object. To make system less
+        /// centralized, and spread the storage + io costs across multiple
+        /// destinations, we can track every `Channel`'s messages.
+        messages: VecSet<vector<u8>>,
```

Comment:
We shouldn't really use VecSet for this purpose as its functionality is rather limiting (+ object size limit). Consider using dfs (expensive) or maybe even shard by the first byte into a set of dfs.

Reply:
Messages can be resubmitted iirc.